runtime/common/src/message_router/mod.rs

```diff
@@ -0,0 +1,263 @@
+	/// This is used to know how to pay for XCM remote execution use local asset.
+	/// For example:
+	/// key: {parents: 1, Parachain(2023)}, val: 14719736222326895902025
+	/// represents the units per second of CRAB token on moonriver
+	#[pallet::storage]
+	#[pallet::getter(fn target_xcm_exec_config)]
+	pub type TargetXcmExecConfig<T: Config> =
+		StorageMap<_, Blake2_128Concat, MultiLocation, AssetUnitsPerSecond>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the units per second of local asset used in target chain.
+		#[pallet::weight(
+			<T as Config>::WeightInfo::set_target_xcm_exec_config()
+		)]
+		pub fn set_target_xcm_exec_config(
+			origin: OriginFor<T>,
+			target_location: MultiLocation,
+			local_asset_units_per_second: AssetUnitsPerSecond,
+		) -> DispatchResultWithPostInfo {
+			T::ConfigModifierOrigin::ensure_origin(origin)?;
+
+			TargetXcmExecConfig::<T>::insert(&target_location, &local_asset_units_per_second);
+
+			Self::deposit_event(Event::TargetXcmExecConfigChanged {
+				target_location,
+				local_asset_units_per_second,
+			});
+
+			Ok(().into())
+		}
+
+		/// Deliver received LCMP messages to other parachains.
+		/// 1. Calculate the fee for xcm remote execution
+		/// 2. Transfer xcm fee to target sovereign account
+		/// 3. Assemble xcm that needs to be executed remotely
+		/// 4. Send xcm to target chain
+		#[pallet::weight(
+			<T as Config>::WeightInfo::forward()
+		)]
+		pub fn forward(
+			origin: OriginFor<T>,
+			target: Target,
+			message: Box<VersionedXcm<<T as frame_system::Config>::Call>>,
+		) -> DispatchResultWithPostInfo {
+			// MultiLocation origin used to execute xcm
+			let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin.clone())?;
+			let account_id = ensure_signed(origin)?;
+			// U8 account used in DescendOrigin instruction
+			let raw_account = <[u8; 32]>::try_from(account_id.encode())
+				.map_err(|_| Error::<T>::AccountIdConversionFailed)?;
+
+			let mut remote_xcm: Xcm<<T as frame_system::Config>::Call> =
+				(*message).try_into().map_err(|()| Error::<T>::BadVersion)?;
+
+			// Calculate the execution fee required for remote xcm execution
+			// fee = fee_per_second * (weight/weight_per_second)
+			let local_asset_units_per_second: AssetUnitsPerSecond;
+			let remote_weight: Weight;
+			match target {
+				Target::Moonbeam => {
+					local_asset_units_per_second =
+						TargetXcmExecConfig::<T>::get(T::MoonbeamLocation::get())
+							.ok_or(Error::<T>::TargetXcmExecNotConfig)?;
+					remote_weight = T::MoonbeamWeigher::weight(&mut Self::extend_remote_xcm(
+						raw_account.clone(),
+						remote_xcm.clone(),
+						MultiAsset { id: AssetId::from(T::LocalAssetId::get()), fun: Fungible(0) },
+					))
+					.map_err(|()| Error::<T>::UnweighableMessage)?;
+				},
```

Comment:
What about the processing of other branches

Reply:
no other branch